### DISCUSSION TITLE
block-map and comments

### DISCUSSION THREAD

**Discussion by revolunet** [eemeli/yaml#490] [Q&A]
Hello, i'm trying to extract comments for each value in such a YAML :

```yaml
# comment 1
create:
  # comment for value 1
  value1:
  # comment for value 2
  value2:
```

When i do 

```js
  const doc = parseDocument(file);
  console.log(doc.contents);
```

I get

```ts
YAMLMap {
    items: [
      Pair {
        key: Scalar {
          value: 'create',
          range: [ 12, 18, 18 ],
          source: 'create',
          type: 'PLAIN',
          commentBefore: ' comment 1'
        },
        value: YAMLMap {
          items: [
            Pair {
              key: Scalar {
                value: 'value1',
                range: [ 46, 52, 52 ],
                source: 'value1',
                type: 'PLAIN'
              },
              value: Scalar {
                value: null,
                range: [ 53, 53, 77 ],
                source: '',
                type: 'PLAIN',
                comment: ' comment for value 2'
              }
            },
            Pair {
              key: Scalar {
                value: 'value2',
                range: [ 80, 86, 86 ],
                source: 'value2',
                type: 'PLAIN',
                spaceBefore: true
              },
              value: Scalar {
                value: null,
                range: [ 87, 87, 87 ],
                source: '',
                type: 'PLAIN'
              }
            }
          ],
          range: [ 46, 87, 87 ],
          commentBefore: ' comment for value 1'
        }
      }
    ],
    range: [ 12, 87, 87 ]
  },
  range: [ 12, 87, 87 ]
}
```

the `create` has the correct `commentBefore` but `value1` and `value2` donc have such `commentBefore`

Not sure if this is intended behaviour or if theres a way to get the `commentBefore` on each value ?

**Comment by eemeli** [comment]
This is intended behaviour. The "comment for value 1" gets assigned to the collection as a whole (not sure, but adding an empty row before it may have an effect), while "comment for value 2" ends up included in the value of the `value1` key -- it's empty, so we need to keep on scanning until we can be sure that there's no value (you could try dedenting it).

YAML does not specify how comments are attached, and as you see here, there isn't always just one way to attach a comment. You're welcome to suggest changes in the logic for comment attachment, but you'll probably need to submit a PR yourself. Be warned, though, the heuristics on this can get a bit hairy as it's pretty much all corner cases.

**Reply by revolunet** [reply]
yes thanks, i'm exploring these cases :)

My goal is to produce a json-schema; Since i have all the data i need here, i "just" have to reorder/remap comments to scalars. How would you iterate that tree in `doc.contents.items` ? Looks like `CST.visit` doesnt fit here

**Reply by eemeli** [reply]
https://eemeli.org/yaml/#modifying-nodes

**Comment by revolunet** [comment]
note: if i use `String(doc)` on the above example, i get this slightly different YAML in return :

```yaml
# comment 1
create:
  # comment for value 1
  value1: # comment for value 2
    
  value2:
 ```
